### PR TITLE
Fix `healAmbiguous` to `compareAlternatives` with `disambiguate = true`

### PR DESCRIPTION
**Pull request by EugeneFlesselle**
Based on #21328
Opened here to have it run the CI

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/typer/Implicits.scala [hunk 1]
```diff
@@ -1439,9 +1439,14 @@ trait Implicits:
...
              *  If no such candidate is found, we propagate the ambiguity.
              */
-            def healAmbiguous(fail: SearchFailure, betterThanFailed: Candidate => Boolean) =
-              val newPending = remaining.filter(betterThanFailed)
-              rank(newPending, fail, Nil).recoverWith(_ => fail)
+            def healAmbiguous(fail: SearchFailure, ambiguousCands: List[RefAndLevel]) =
+              def betterThanAll(newCand: RefAndLevel, disambiguate: Boolean): Boolean =
+                ambiguousCands.forall(compareAlternatives(newCand, _, disambiguate) > 0)
+
+              val newPending = remaining.filter(betterThanAll(_, disambiguate = false))
```
Comment (odersky) [inline compiler/src/dotty/tools/dotc/typer/Implicits.scala:1446]: This looks like it does the right thing. But it does cost extra work, including in source versions where we don't even compare old and new implicits. Can we avoid that?

Comment (EugeneFlesselle) [reply compiler/src/dotty/tools/dotc/typer/Implicits.scala:1446]: Yes indeed, the check should be redundant if `!isPriorityChangeVersion`

I'll update that

Comment (odersky) [reply compiler/src/dotty/tools/dotc/typer/Implicits.scala:1446]: But even then we run the compareAlternatives twice...

Comment (EugeneFlesselle) [reply compiler/src/dotty/tools/dotc/typer/Implicits.scala:1446]: An alternative would to keep track of the nesting level in `SearchFailureType`. We would then be able to test between the possible origins of `healAmbiguous` in the case SearchFailure of `disambiguate`

**Review by odersky — APPROVED** [review]
Looks all good to me now

**Comment by odersky** [comment]
@EugeneFlesselle Please merge unless there's something else you want to add.

**Pull request merged**